floors/enforce.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return seatBids, []error{errors.New("Error in getting request extension")}, rejectedBids
+	}
+
+	if isPriceFloorsDisabled(account, bidRequestWrapper) {
+		return seatBids, []error{errors.New("Floors feature is disabled at account or in the request")}, rejectedBids
+	}
+
+	if !isFloorsSignallingSkipped(requestExt) && isValidImpBidFloorPresent(bidRequestWrapper.BidRequest) {
+		if enforceFloors := isSatisfiedByEnforceRate(requestExt, account.PriceFloors.EnforceFloorsRate, rand.Intn); enforceFloors {
+			enforceDealFloors := account.PriceFloors.EnforceDealFloors && getEnforceDealsFlag(requestExt)
+
+			seatBids, rejectionErrs, rejectedBids = enforceFloorToBids(bidRequestWrapper, seatBids, conversions, enforceDealFloors)
+		}
+	}
+	return seatBids, rejectionErrs, rejectedBids
+}
+
+// enforceFloorToBids function does floors enforcement for each bid,
+// The bids returned by each partner below bid floor price are rejected and remaining eligible bids are considered for further processing
+func enforceFloorToBids(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, conversions currency.Conversions, enforceDealFloors bool) (map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, []error, []*entities.PbsOrtbSeatBid) {
+	errs := []error{}
+	rejectedBids := []*entities.PbsOrtbSeatBid{}
+	impMap := make(map[string]*openrtb_ext.ImpWrapper, bidRequestWrapper.LenImp())
+
+	for _, v := range bidRequestWrapper.GetImp() {
+		impMap[v.ID] = v
+	}
+
+	for bidderName, seatBid := range seatBids {
+		eligibleBids := make([]*entities.PbsOrtbBid, 0, len(seatBid.Bids))
+		for _, bid := range seatBid.Bids {
+
+			reqImp, ok := impMap[bid.Bid.ImpID]
+			if !ok {
+				continue
+			}
+
+			reqImpCur := reqImp.BidFloorCur
+			if reqImpCur == "" {
+				reqImpCur = defaultCurrency
+				if bidRequestWrapper.Cur != nil {
+					reqImpCur = bidRequestWrapper.Cur[0]
+				}
+			}
+			updateBidExtWithFloors(reqImp, bid, reqImpCur)
+
+			if !isPriceFloorsEnforcementDisabled(bidRequestWrapper) {
+				dealBid := checkDealBidForEnforcement(bid, enforceDealFloors)
+				if dealBid != nil {
+					eligibleBids = append(eligibleBids, dealBid)
+					continue
+				}
+
+				rate, err := getCurrencyConversionRate(seatBid.Currency, reqImpCur, conversions)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("error in rate conversion from = %s to %s with bidder %s for impression id %s and bid id %s error = %v", seatBid.Currency, reqImpCur, bidderName, bid.Bid.ImpID, bid.Bid.ID, err.Error()))
+					continue
+				}
+
+				bidPrice := rate * bid.Bid.Price
+
+				if reqImp.BidFloor > bidPrice {
+					rejectedBid := &entities.PbsOrtbSeatBid{
+						Currency: seatBid.Currency,
+						Seat:     seatBid.Seat,
+						Bids:     []*entities.PbsOrtbBid{bid},
+					}
+					rejectedBids = append(rejectedBids, rejectedBid)
+					continue
+				}
+			}
+			eligibleBids = append(eligibleBids, bid)
```

Comment:
Nitpick: instead of using `continue` here can we just put the `eligibleBids` assignment in an else block since this logic is part of the normal flow (no early return)?

Reply:
Done